hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java

```diff
@@ -1612,6 +1613,20 @@ private boolean isOpenContainerHealthy(
         .allMatch(r -> compareState(state, r.getState()));
   }
 
+  private void setHealthStateForClosing(Set<ContainerReplica> replicas,
+                                        ContainerInfo container,
+                                        ReplicationManagerReport report) {
+    if (!replicas.stream().
+            anyMatch(r -> compareState(LifeCycleState.OPEN, r.getState()) ||
+                    compareState(LifeCycleState.CLOSED, r.getState()))) {
+      report.incrementAndSample(HealthState.MISSING, container.containerID());
```

Comment:
The only way for a container to be "MISSING" is if there are no replicas at all, as otherwise there is a replica available to read. So can we not just say `if (replicas.size() == 0) ...` ?

Reply:
It is not enough. For example, lets say we have 6 data nodes and replication factor 3. Container #1 can be stored on data nodes 1, 2 and 4, for example. If all 3 data nodes (data nodes 1, 2 and 4) go offline at the same moment, than the container #1 will be in CLOSING state and will have 2 replicas also in CLOSING state but it will still be MISSING. 
I am new on this project, so I might be wrong.

Reply:
The definition of MISSING is that the container is not able to be read because no replicas exist on any online nodes. If you have 3 replicas and then all 3 DNs go offline, then there will be zero replicas available. The replicas in SCM will be reduced to zero and hence `replicas.size() == 0` will be true. When a DN goes offline and is marked DEAD (after 10 minutes) its replicas are removed from SCM automatically - it does not remember the last location. If the node comes back online, then the node will sent a replica report which will add the replica location back to SCM.